### PR TITLE
Log only to the explicitly-requested destination

### PR DESCRIPTION
**Pull request by quodlibetor**
If users specify a --log-file we should assume that they know what they're
doing and write logs to exactly that destination, and nowhere else.

This removes the secondary stderr `WARN` logs that were impossible to silence,
if users provide a `--log-file` (that is not stderr).

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4777)

### CODE REVIEW

**Review by cirego — APPROVED** [review]

File: src/materialized/src/bin/materialized/main.rs [hunk 1]
```diff
@@ -323,55 +323,66 @@ fn run() -> Result<(), anyhow::Error> {
...
-                })
-                .with(FilterLayer::new(
-                    fmt::layer().with_writer(io::stderr),
-                    LevelFilter::WARN,
-                ))
-                .init();
+            {
+                tracing_subscriber::registry()
+                    .with(env_filter)
+                    .with(fmt::layer().with_writer(io::stderr))
```
Comment (cirego) [inline src/materialized/src/bin/materialized/main.rs:343]: Why is this `stderr` by default and not `stdout`? Do any logs go to `stdout`?

**Comment by benesch** [comment]
What is the problem that’s being solved here? I agree that logging is not thoroughly configurable at the moment, but I’m not convinced that this is the right solution.

**Comment by quodlibetor** [comment]
> What is the problem that’s being solved here?

Well-behaved systems should do exactly what you tell them, and appropriate behavior should be configurable. In this case, running with a `--log-file` parameter in a startup/systemd script still requires redirecting stderr to /dev/null. Users who know what their doing well enough to configure a log file should expect that only that log file is written to.

I agree that generally configurable logs are better, but this seems like a strict improvement over the status quo. 

What is your concern with this solution?

**Comment by benesch** [comment]
A couple things, off the top of my head:

  * Logging warnings and up to stderr ensures that panic messages are printed to both stderr and the log file, which has been important in the past.
  * I don't think the code here matches the PR description. Based on the PR description, I would expect a bare `materialized` with no explicit `--log-file` to continue to log warnings and up to stderr but other messages to mzdata/materialized.log. But instead I believe it logs nothing to stderr, judging by the fact that the code I wrote to support this, in tracing.rs, has been removed.

I'm still not totally clear on the motivation here. Why do you need to redirect stderr in a systemd unit file? Because the default location for stderr is too slow? I'd contend instead that we are abusing WARN logs. Possibly another fix is to limit what gets sent to stderr to only errors.

I guess what I'm trying to say is that the current design is intentional! If there are deficiencies in the behavior we should definitely correct them, but I think it would be good to discuss those deficiencies on Slack before diving into a particular solution.

**Comment by quodlibetor** [comment]
> I don't think the code here matches the PR description.

You're correct, that's a bug in this PR, I'll update it to what you expect.

File: src/materialized/src/bin/materialized/main.rs [hunk 2]
```diff
@@ -335,43 +335,62 @@ fn run() -> Result<(), anyhow::Error> {
...
+                        fmt::layer().with_ansi(false).with_writer(move || {
+                            file.try_clone().expect("failed to clone log file")
+                        }),
+                    )
+                    .with(FilterLayer::new(
+                        fmt::layer().with_writer(io::stderr),
+                        LevelFilter::WARN,
+                    ))
+                    .init();
+            }
```
Comment (benesch) [inline src/materialized/src/bin/materialized/main.rs:393]: The code here is fiddly enough it'd be great if we could push the conditional down a bit to share more code:

```rust
            Some("stderr") => {
                // The user explicitly directed logs to stderr. Log only to stderr
                // with the user-specified `env_filter`.
                tracing_subscriber::registry()
                    .with(env_filter)
                    .with(fmt::layer().with_writer(io::stderr))
                    .init()
            }
            log_file => {
                // Logging to a file. If the user did not explicitly specify
                // a file, bubble up warnings and errors to stderr.
                let stderr_level = match log_file {
                    Some(_) => LevelFilter::OFF,
                    None => LevelFilter::WARN,
                };
                tracing_subscriber::registry()
                    .with(env_filter)
                    .with({
                        let path = match log_file {
                            Some(log_file) => PathBuf::from(log_file),
                            None => data_directory.join("materialized.log"),
                        };
                        if let Some(parent) = path.parent() {
                            fs::create_dir_all(parent).with_context(|| {
                                format!("creating log file directory: {}", parent.display())
                            })?;
                        }
                        let file = fs::OpenOptions::new()
                            .append(true)
                            .create(true)
                            .open(&path)
                            .with_context(|| format!("creating log file: {}", path.display()))?;
                        fmt::layer().with_ansi(false).with_writer(move || {
                            file.try_clone().expect("failed to clone log file")
                        })
                    })
                    .with(FilterLayer::new(
                        fmt::layer().with_writer(io::stderr),
                        stderr_level,
                    ))
                    .init()
            }
```

File: doc/user/content/release-notes.md [hunk 3]
```diff
@@ -56,6 +56,8 @@ Wrap your release notes at the 80 character mark.
 - Report an error without crashing when a query contains unexpected UTF-8
   characters, e.g., `SELECT ’1’` {{% gh 4755 %}}.
 - [`TAIL`](/sql/tail) now supports timestamp progress.
+- Log messages are no longer emit to stderr if an explicit `--log-file` is
+  supplied at startup. {{ gh 4777 }}
```
Comment (benesch) [inline doc/user/content/release-notes.md:60]: "emitted"

File: src/materialized/src/bin/materialized/main.rs (hunk 1 above)
Comment (quodlibetor) [reply src/materialized/src/bin/materialized/main.rs:344]: no logs go to stdout, all output intended for humans goes to stderr. Right now that means that all output goes to stderr, I believe.

File: src/materialized/src/bin/materialized/main.rs (hunk 2 above)
Comment (quodlibetor) [reply src/materialized/src/bin/materialized/main.rs:380]: done.

**Review by benesch — APPROVED** [review]
:shipit:

**Pull request merged**